### ISSUE TITLE
Regression? Unexpected initial data while revalidating after a local mutation

### ISSUE THREAD

**Issue by JulianG** [zeit/swr#344]
I think have found different behaviour between v0.1.17 and v0.1.18. But maybe I'm using it wrong.

![0117](https://user-images.githubusercontent.com/237818/79193411-7c712880-7e22-11ea-84e0-fb94b87f64a7.gif) ![0118](https://user-images.githubusercontent.com/237818/79193562-c65a0e80-7e22-11ea-88d3-f139bed559fd.gif)

Repro steps:

1. Try the [v0.1.17](https://codesandbox.io/s/swr-0117-mutate-m57ec?file=/src/App.js) codesandbox and press the button. See how the data transitions from "Zero" to "One! (local mutation)" to finally "One!".

2. Try the [v0.1.18](https://codesandbox.io/s/swr-0118-mutate-nipcc?file=/src/App.js) codesandbox and press the button. See how the data transitions from "Zero" to "One! (local mutation)", **then back to "Zero"**, then finally to "One!".

The same issue happens on 0.2.0 and 0.2.1-beta.

I'm not sure what the expected behaviour is. I wasn't expecting seeing the "Zero" value again after having mutated it locally.

This is how I'm mutating: (maybe I'm not supposed to do this?)

```ts
  const changeStatus = async () => {
    // 1. Mutate local cache
    mutate("data", { status: "One! (local mutation)" });
    // 2. call async function to change remote data (and await)
    await setData({ status: "One!" });
    // 3. revalidate
    revalidate();
  };
```

The entire React component below:

```ts
export default function App() {
  const { data, revalidate } = useSWR("data", getData);

  const changeStatus = async () => {
    mutate("data", { status: "One! (local mutation)" });
    await setData({ status: "One!" });
    revalidate();
  };

  return (
    <>
      <h1>SWR v0.1.17/18</h1>
      <pre>{JSON.stringify(data, null, 2)}</pre>
      {data && <button onClick={changeStatus}>Change to One</button>}
    </>
  );
}
```

For reference, below are the `getData` and `setData` async functions:

```ts
let data = {
  status: "Zero"
};

async function getData() {
  await wait(500);
  return data;
}

async function setData(partialData) {
  await wait(500);
  data = { ...data, ...partialData };
}
```

**Comment by JulianG** [comment]
I managed to write a failing test in **use-swr.test.tsx**

```ts
it('should mutate the cache and revalidate in async - issue?', async () => {
  cache.clear()

  function wait(t) {
    return new Promise(resolve => {
      setTimeout(resolve, t)
    })
  }
  let data = 'initial'

  async function getData() {
    await wait(5)
    return data
  }

  async function setData(d: string) {
    await wait(5)
    data = d
  }

  function Page() {
    const { data } = useSWR('dynamic-9', getData, { dedupingInterval: 0 })
    return <div>data: {data}</div>
  }
  const { container } = render(<Page />)

  // hydration
  expect(container.firstChild.textContent).toMatchInlineSnapshot(`"data: "`)
  await waitForDomChange({ container }) // mount
  expect(container.firstChild.textContent).toMatchInlineSnapshot(
    `"data: initial"`
  )
  await act(() => {
    // mutate and revalidate
    mutate('dynamic-9', 'local')
    setData('new-value')
    return new Promise(res => setTimeout(res, 1))
  })
  expect(container.firstChild.textContent).toMatchInlineSnapshot(
    `"data: local"`
  )

  await act(() => new Promise(res => setTimeout(res, 200))) // recovers
  expect(container.firstChild.textContent).toMatchInlineSnapshot(
    `"data: new-value"` // test fails here because data is `initial`
  )
})
```

**Comment by sergiodxa** [comment]
This is happening because your mutate call is making SWR call `getData` before you call `setData` and is finishing earlier.

```js
export default function App() {
  const { data, revalidate } = useSWR("data", getData);

  const changeStatus = async () => {
    // the false here ensure SWR only mutate without calling revalidate
    mutate("data", { status: "One! (local mutation)" }, false);
    await setData({ status: "One!" });
    revalidate();
  };

  return (
    <>
      <h1>SWR v0.1.17/18</h1>
      <pre>{JSON.stringify(data, null, 2)}</pre>
      {data && <button onClick={changeStatus}>Change to One</button>}
    </>
  );
}
```

I tried in your second CodeSandbox and this way it works. I think the behavior without the `false` is the expected behavior, the idea here is that you can mutate locally and get SWR revalidate immediately.

Another option here is to do:

```js
async function setData(partialData) {
  await wait(900);
  data = { ...data, ...partialData };
  return data; // be sure to return something here
}

export default function App() {
  const { data } = useSWR("data", getData);

  const changeStatus = async () => {
    // the false here ensure SWR only mutate without calling revalidate
    mutate("data", { status: "One! (local mutation)" }, false);
    // here we update the data and use the returned value as new cached value
    // we still let mutate revalidate against the original fetcher function to ensure
    // we have it correctly, you can disable it if you trust the response is correct
    mutate("data", setData({ status: "One!" }));
  };

  return (
    <>
      <h1>SWR v0.1.17/18</h1>
      <pre>{JSON.stringify(data, null, 2)}</pre>
      {data && <button onClick={changeStatus}>Change to One</button>}
    </>
  );
}
```

This way you only do two lines.

**Comment by JulianG** [comment]
OH! you're right. I guess it was a case of RTFM. Thanks for taking the time to answer!

I guess the default was 'false' in v0.1.17 ?

**Comment by sergiodxa** [comment]
I think it's because of this `https://github.com/zeit/swr/issues/271` fix that was included in `0.1.18`, also note current version is `0.2.0`.

**Comment by JulianG** [comment]
Thanks! Yes. I found the "issue" on 0.2.0, but then narrowed it down to 0.1.18.
Well, it's not an issue anyways! 😄
